File: 2020-04/4-10/21合并两个有序链表/21.py
#  -*- coding: utf-8 -*-

#  Editor      : Pycharm
#  File        : 21.py
#  Author      : ZhenyuMa
#  Created     : 2020/4/10 11:03 下午
#  Description : easy


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        cur1 = l1
        cur2 = l2
        dummy_head = ListNode(-1)
        pre = dummy_head

        while cur1 or cur2:
            if not cur1:
                pre.next = cur2
                cur2 = cur2.next
            elif not cur2:
                pre.next = cur1
                cur1 = cur1.next
            elif cur1.val <= cur2.val:
                pre.next = cur1
                cur1 = cur1.next
            else:
                pre.next = cur2
                cur2 = cur2.next
            pre = pre.next

        return dummy_head.next

File: 2020-05/5-01/21合并连个有序链表/21.cpp

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummyhead = new ListNode(-1);
        ListNode* pre = dummyhead;

        while(l1 != nullptr && l2 != nullptr){
            if(l1->val <= l2->val){
                pre->next = l1;
                l1 = l1->next;
            }
            else{
                pre->next = l2;
                l2 = l2->next;
            }
            pre = pre->next;
        }
        if(l1 != nullptr)
            pre->next = l1;
        else
            pre->next = l2;
        return dummyhead->next;
    }
};